tests/playwright-test/test-server-connection.spec.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { test as baseTest, expect } from './ui-mode-fixtures';
+
+import { TestServerConnection } from '../../packages/playwright/src/isomorphic/testServerConnection';
+
+class TestServerConnectionUnderTest extends TestServerConnection {
+  events: [string, any][] = [];
+
+  constructor(wsUrl: string) {
+    super(wsUrl);
+    this.onTestFilesChanged(params => this.events.push(['testFilesChanged', params]));
+    this.onStdio(params => this.events.push(['stdio', params]));
+    this.onLoadTraceRequested(params => this.events.push(['loadTraceRequested', params]));
+    this.onReport(params => this.events.push(['report', params]));
+  }
+}
+
+const test = baseTest.extend<{ testServerConnection: TestServerConnectionUnderTest }>({
+  testServerConnection: async ({ startCLICommand }, use) => {
+    const testServerProcess = await startCLICommand({}, 'test-server');
+
+    await testServerProcess.waitForOutput('Listening on');
+    const line = testServerProcess.output.split('\n').find(l => l.includes('Listening on'));
+    const wsEndpoint = line!.split(' ')[2];
+
+    await use(new TestServerConnectionUnderTest(wsEndpoint));
+
+    await testServerProcess.kill();
+  }
+});
+
+test('file watching', async ({ testServerConnection, writeFiles }, testInfo) => {
+  await writeFiles({
+    'utils.ts': `
+      export const expected = 42;
+      `,
+    'a.test.ts': `
+      import { test } from '@playwright/test';
+      import { expected } from "./utils";
+      test('foo', () => {
+        expect(123).toBe(expected);
+      });
+      `,
+  });
+
+  const tests = await testServerConnection.listTests({});
+  expect(tests.report.map(e => e.method)).toEqual(['onConfigure', 'onProject', 'onBegin', 'onEnd']);
+
+  await testServerConnection.watch({ fileNames: ['a.test.ts'] });
+
+  await writeFiles({
+    'utils.ts': `
+      export const expected = 123;
+      `,
+  });
+
+  await expect.poll(() => testServerConnection.events).toHaveLength(1);
+  expect(testServerConnection.events).toEqual([
+    ['testFilesChanged', { testFiles: [testInfo.outputPath('a.test.ts')] }]
```

Comment:
Are we trying to test that there is a single event here? Seems kinds of redundant, because we rely on chokidar to make it so.

Reply:
We're trying to test that there are no other events. Prior to this PR, there would've also been a `listChanged` event.

Also, prior to this PR, we would've seen two `testFilesChanged` events because we had two chokidar instances emitting the same file changes. So there's some value in this test :D If it had existed before, we would've spotted earlier that the two `Watcher` instances are a redundant.